### PR TITLE
Team Flare Kalos Quest Storyline

### PR DESCRIPTION
**Pull request by LuchoDreamer**
This adds a quest for Team Flare, the villain team of Kalos. Following the story of the X/Y games.

The quest starts after you beat Sycamore 1 in Lumiose City, and it ends with a temporary fight after you become champion.

The quest rewards a random Kalos fossil around the start, a master ball midway, and a Key Stone at the end. The quest also unlocks the fight against AZ for Eternal Floette (instead of being on champion), and it unlocks a fight against Team Flare Boss Lysandre in Couriway Town that gives the Gyaradosite. On top of that, this temp battle only appears during dusk, symbolizing how in the games the Gyaradosite only appears from 8pm to 9pm.

This quest changes a lot of requirements for unlocking cities/gyms/dungeons, so old players will have to beat the quest to get access to some places again. It also changes the requirement for Xerneas and Yveltal, which are tied to the quest, and the requirements of Abomasnow and Volcanion so you need specific steps of the quest done to be available again so the quest makes sense.

ETA: I forgot to mention, this PR also adds DayCyclePartRequirement on Temporary Battles, so a requirement for a part of the day can be added to them.

### CODE REVIEW

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 1]
```diff
@@ -2754,7 +2885,7 @@ TemporaryBattleList['Grand Duchess Diantha'] = new TemporaryBattle(
         new GymPokemon('Goodra', 120770015, 80),
         new GymPokemon('Mega Gardevoir', 125000000, 80),
     ],
-    'Witnessing the noble spirits of you and your Pokémon in battle has really touched my heart... Please, take this Gardevoirite.</br><img src="assets/images/megaStone/282.png"/></br>I just... I just don\'t know what to say... I can hardly express this feeling...',
+    'Witnessing the noble spirits of you and your Pokémon in battle has really touched my heart... Please, take this Gardevoirite.</br><img src="assets/images/megaStone/130.png"/></br>I just... I just don\'t know what to say... I can hardly express this feeling...',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2888]: Not sure why you changed this? Garvevoir's ID is  282, not 130

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 2]
```diff
@@ -2764,6 +2895,25 @@ TemporaryBattleList['Grand Duchess Diantha'] = new TemporaryBattle(
...
+        new GymPokemon('Murkrow', 127138249, 79),
+        new GymPokemon('Pyroar', 124595484, 81),
+        new GymPokemon('Mega Gyarados', 129681014, 83),
+    ],
+    'I see the strength to protect within you... Hopefully you can do a better job than me and protect a tomorrow that will be better than today. Take this Gyaradosite.</br><img src="assets/images/megaStone/130.png"/></br>',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 45), new ObtainedPokemonRequirement('Gyarados'), new DayCyclePartRequirement([DayCyclePart.Dusk])],//new DayCyclePartRequirement([DayCyclePart.Dusk])
+    undefined,
+    {
+        displayName: 'Team Flare Boss Lysandre',
+        returnTown: 'Couriway Town',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2911]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 3]
```diff
@@ -2463,6 +2496,81 @@ TemporaryBattleList['Calem 4'] = new TemporaryBattle(
...
+        new GymPokemon('Murkrow', 68927443, 49),
+        new GymPokemon('Pyroar', 70334126, 51),
+        new GymPokemon('Mega Gyarados', 71740808, 53),
+    ],
+    'It\'s over... Fools with no vision will continue to befoul this beautiful world. They will go on until the only thing left to do is squabble over the remaining scraps of hope...',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 42)],
+    undefined,
+    {
+        displayName: 'Team Flare Boss Lysandre',
+        returnTown: 'Team Flare Secret HQ',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2570]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 4]
```diff
@@ -2463,6 +2496,81 @@ TemporaryBattleList['Calem 4'] = new TemporaryBattle(
...
+);
+TemporaryBattleList.Yveltal = new TemporaryBattle(
+    'Yveltal',
+    [new GymPokemon('Yveltal', 281500000, 50)],
+    'You calmed down Yveltal and it ran away to the back of the lab.',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 40)],
+    undefined,
+    {
+        displayName: 'Yveltal',
+        returnTown: 'Team Flare Secret HQ',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2552]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 5]
```diff
@@ -2463,6 +2496,81 @@ TemporaryBattleList['Calem 4'] = new TemporaryBattle(
...
+);
+TemporaryBattleList.Xerneas = new TemporaryBattle(
+    'Xerneas',
+    [new GymPokemon('Xerneas', 281500000, 50)],
+    'You calmed down Xerneas and it ran away to the back of the lab.',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 40)],
+    undefined,
+    {
+        displayName: 'Xerneas',
+        returnTown: 'Team Flare Secret HQ',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2539]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 6]
```diff
@@ -2463,6 +2496,81 @@ TemporaryBattleList['Calem 4'] = new TemporaryBattle(
...
+    [
+        new GymPokemon('Crobat', 128657279, 46),
+        new GymPokemon('Malamar', 131365853, 48),
+    ],
+    'Oh ho ho! You\'re wonderful! Your Pokémon are amazing.',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 28)],
+    undefined,
+    {
+        displayName: 'Team Flare Xerosic',
+        returnTown: 'Lumiose City',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2527]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 7]
```diff
@@ -2463,6 +2496,81 @@ TemporaryBattleList['Calem 4'] = new TemporaryBattle(
...
+        new GymPokemon('Murkrow', 66360070, 45),
+        new GymPokemon('Pyroar', 67714357, 47),
+        new GymPokemon('Gyarados', 69068644, 49),
+    ],
+    'Excellent! I can feel the fire of your convictions burning deep within your heart!',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 25)],
+    undefined,
+    {
+        displayName: 'Team Flare Lysandre',
+        returnTown: 'Lumiose City',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:2512]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 8]
```diff
@@ -1917,6 +1920,36 @@ TemporaryBattleList['Trevor & Tierno'] = new TemporaryBattle(
...
+    [
+        new GymPokemon('Houndour', 37260908, 21),
+        new GymPokemon('Golbat', 39069360, 23),
+    ],
+    'You may have beaten me... again, but when I lose, I go out in style!',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 8)],
+    undefined,
+    {
+        displayName: 'Team Flare Grunt',
+        returnTown: 'Glittering Cave',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:1949]: Doesn't need a returnTown

File: src/scripts/temporaryBattle/TemporaryBattleList.ts [hunk 9]
```diff
@@ -1917,6 +1920,36 @@ TemporaryBattleList['Trevor & Tierno'] = new TemporaryBattle(
...
+    'Team Flare Grunt 1',
+    [
+        new GymPokemon('Houndour', 35486579, 18),
+        new GymPokemon('Zubat', 36513421, 18),
+    ],
+    'You may have beaten me, but when I lose, I go out in style!',
+    [new QuestLineStepCompletedRequirement('A Beautiful World', 4)],
+    undefined,
+    {
+        displayName: 'Team Flare Grunt',
```
Comment (Qwertypop04) [inline src/scripts/temporaryBattle/TemporaryBattleList.ts:1933]: Doesn't need a returnTown

File: src/scripts/towns/TownList.ts [hunk 10]
```diff
@@ -4708,12 +4708,77 @@ const LumioseDexio = new NPC('Dexio', [
...
+    requirement: new MultiRequirement([new QuestLineStepCompletedRequirement('A Beautiful World', 7), new QuestLineStepCompletedRequirement('A Beautiful World', 9, GameConstants.AchievementOption.less)]),
+});
+
+const TeamFlareGrunt3 = new NPC('Team Flare Grunt', [
+    'Hear this, pip-squeak! Team Flare is gonna use the stones on Route 10 and a fantastic treasure to make everyone on our team happy!',
+    'In this world the ones who have power are the ones who win with flair!',
+
+], {
+    image: 'assets/images/trainers/Team Flare Grunt (male).png',
+    requirement: new MultiRequirement([new QuestLineStepCompletedRequirement('A Beautiful World', 9), new QuestLineStepCompletedRequirement('A Beautiful World', 10, GameConstants.AchievementOption.less)]),
```
Comment (Qwertypop04) [inline src/scripts/towns/TownList.ts:4779]: The disappear requirement here should be set to 11, so it doesn't disappear instantly upon being read

Comment (LuchoDreamer) [reply src/scripts/towns/TownList.ts:4779]: This one is on purpose. It's supposed to disappear instantly after the conversation and you cannot "follow" him.

**Review by Jaaslet — APPROVED** [review]
Looks good. Thanks for the PR

**Pull request merged**